### PR TITLE
Retry upsert of command deduplication on oracle and h2 [DPP-609]

### PR DESCRIPTION
**Pull request by mziolekda**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 1]
```diff
@@ -91,14 +93,26 @@ private[backend] object H2StorageBackend
...
-      .executeUpdate()(connection)
+  )(connection: Connection): Int = {
+    // When a deduplication upsert is performed simultaneously from multiple threads, the query fails with
+    // JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation
+    // Simple retry helps
+    def retry[T](op: => T): T =
+      try {
+        op
+      } catch {
+        case NonFatal(_) => op
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:104]: Should we log (and debug/trace level) that a retry took place and what was the cause?

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 2]
```diff
@@ -91,14 +93,26 @@ private[backend] object H2StorageBackend
...
-    SQL(SQL_INSERT_COMMAND)
-      .on(
-        "deduplicationKey" -> key,
-        "submittedAt" -> Timestamp.instantToMicros(submittedAt),
-        "deduplicateUntil" -> Timestamp.instantToMicros(deduplicateUntil),
-      )
-      .executeUpdate()(connection)
+  )(connection: Connection): Int = {
+    // When a deduplication upsert is performed simultaneously from multiple threads, the query fails with
+    // JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:98]: Should we also mention under what settings this problem occurs?
In particular I have in mind the enabled transaction isolation level: http://www.h2database.com/html/advanced.html#transaction_isolation.
If we were to change the isolation level we'll then have some context to opine if this retry is still needed.

**Comment by oliverse-da** [comment]
Is this the PR we are wanted to port to the `release/1.17.0` branch? (If not please clear the `needs-backport` label that I just stuck on this PR)

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala (hunk 2 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:98]: good point

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala (hunk 1 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:104]: I was thinking of logging at info, which seemed an overkill. I could go with a debug message.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 3]
```diff
@@ -91,14 +95,30 @@ private[backend] object H2StorageBackend
...
+    // Under the default READ_COMMITTED isolation level used for the indexdb, when a deduplication
+    // upsert is performed simultaneously from multiple threads, the query fails with
+    // JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation
+    // Simple retry helps
+    def retry[T](op: => T): T =
+      try {
+        op
+      } catch {
+        case NonFatal(e) =>
+          logger.debug(s"Caught exception while upsering a deduplication entry: $e")
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:109]:
```suggestion
          logger.debug(s"Caught exception while upserting a deduplication entry: $e")
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 4]
```diff
@@ -91,14 +95,30 @@ private[backend] object H2StorageBackend
...
+
+    // Under the default READ_COMMITTED isolation level used for the indexdb, when a deduplication
+    // upsert is performed simultaneously from multiple threads, the query fails with
+    // JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation
+    // Simple retry helps
+    def retry[T](op: => T): T =
+      try {
+        op
+      } catch {
+        case NonFatal(e) =>
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:108]: Is a single retry without a randomized back-off time enough to overcome potentially multiple conflicts? Or do we consider the risk of this happening reduced enough to accept more rare occasions?

Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:108]: The conflict can happen only against a single winner. Once that's settled retry will see the proper committed result.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala (hunk 3 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala]: :facepalm:

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala (hunk 4 above)
Comment (gerolf-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:108]: Of course, thank you.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala (hunk 3 above)
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala]: Minor: Consider passing the throwable instance as the second parameter to the debug method.

**Review by pbatko-da — APPROVED** [review]

**Review by rautenrieth-da — COMMENTED** [review]
I think we should add a `StorageBackendTestsDeduplication` for this. We do test command deduplication in the ledger api test tool (`CommandDeduplicationBase`), but going through the ledger API means it's more difficult to generate conflicting upsert queries, and we have no control over when expired entries are garbage-collected through `StorageBackend.removeExpiredDeduplicationData`.

Something alone the lines of:
1. Call `StorageBackend.upsertDeduplicationEntry` N times in parallel, and verify that exactly one of them returns 1 (exactly one query inserts a new row).
2. Insert an expired deduplication entry, then call `StorageBackend.upsertDeduplicationEntry` N times in parallel, and verify that exactly one of them returns 1 (exactly one query updates the existing row).

**Pull request merged**